### PR TITLE
MTDSA-14152: TYS Updates For Create And Amend Relief Investments

### PR DESCRIPTION
**Pull request by tom-lloyd-hmrc**

### CODE REVIEW

File: app/v1/controllers/CreateAndAmendReliefInvestmentsController.scala [hunk 1]
```diff
@@ -106,10 +106,19 @@ class CreateAndAmendReliefInvestmentsController @Inject() (val authService: Enro
...
 
     errorWrapper.error match {
-      case NinoFormatError | BadRequestError | TaxYearFormatError | RuleIncorrectOrEmptyBodyError | RuleTaxYearNotSupportedError |
-          RuleTaxYearRangeInvalidError | MtdErrorWithCustomMessage(ValueFormatError.code) | MtdErrorWithCustomMessage(
-            DateOfInvestmentFormatError.code) | MtdErrorWithCustomMessage(NameFormatError.code) | MtdErrorWithCustomMessage(
-            UniqueInvestmentRefFormatError.code) =>
+      case _
+          if errorWrapper.containsAnyOf(
+            NinoFormatError,
+            BadRequestError,
```
Comment (sean-flynn-HMRC) [inline app/v1/controllers/CreateAndAmendReliefInvestmentsController.scala:112]: Curious as to why we tend to include "BadRequestError" in the errorWrapper mappings. It's not included in the specs so surely all required error scenarios are covered by the other errors?

Comment (tom-lloyd-hmrc) [reply app/v1/controllers/CreateAndAmendReliefInvestmentsController.scala:112]: The [requestParser](https://github.com/hmrc/individuals-reliefs-api/blob/main/app/v1/controllers/requestParsers/RequestParser.scala#L40) class can return a BadRequestError in the case where there a multiple validation errors.

Comment (tom-lloyd-hmrc) [reply app/v1/controllers/CreateAndAmendReliefInvestmentsController.scala:112]: It's in the spec under the `Multiple validation type error` section.

File: app/v1/services/CreateAndAmendReliefInvestmentsService.scala [hunk 2]
```diff
@@ -40,19 +39,25 @@ class CreateAndAmendReliefInvestmentsService @Inject() (connector: CreateAndAmen
...
-    Map(
+  private def downstreamErrorMap = {
+    val errors = Map(
       "INVALID_TAXABLE_ENTITY_ID" -> NinoFormatError,
-      "FORMAT_TAX_YEAR"           -> TaxYearFormatError,
+      "INVALID_TAX_YEAR"          -> TaxYearFormatError,
+      "INVALID_PAYLOAD"           -> InternalError,
+      "UNPROCESSABLE_ENTITY"      -> InternalError,
       "SERVER_ERROR"              -> InternalError,
       "SERVICE_UNAVAILABLE"       -> InternalError
```
Comment (sean-flynn-HMRC) [inline app/v1/services/CreateAndAmendReliefInvestmentsService.scala:52]: I think the old tech spec may have missed out an INVALID_CORRELATIONID error (i..e no second underscore _). The downstream spec for #1629 does include a INVALID_CORRELATIONID error so probs best to add a mapping for it here and update the old spec too
image

Comment (tom-lloyd-hmrc) [reply app/v1/services/CreateAndAmendReliefInvestmentsService.scala:52]: Good spot!

File: it/v1/endpoints/CreateAndAmendReliefInvestmentsControllerISpec.scala [hunk 3]
```diff
@@ -694,14 +521,68 @@ class CreateAndAmendReliefInvestmentsControllerISpec extends IntegrationBaseSpec
...
 
-        val input = Seq(
+        val errors = List(
           (BAD_REQUEST, "INVALID_TAXABLE_ENTITY_ID", BAD_REQUEST, NinoFormatError),
-          (BAD_REQUEST, "FORMAT_TAX_YEAR", BAD_REQUEST, TaxYearFormatError),
+          (BAD_REQUEST, "INVALID_TAX_YEAR", BAD_REQUEST, TaxYearFormatError),
+          (BAD_REQUEST, "INVALID_PAYLOAD", INTERNAL_SERVER_ERROR, InternalError),
+          (UNPROCESSABLE_ENTITY, "UNPROCESSABLE_ENTITY", INTERNAL_SERVER_ERROR, InternalError),
           (SERVICE_UNAVAILABLE, "SERVICE_UNAVAILABLE", INTERNAL_SERVER_ERROR, InternalError),
           (INTERNAL_SERVER_ERROR, "SERVER_ERROR", INTERNAL_SERVER_ERROR, InternalError)
```
Comment (sean-flynn-HMRC) [inline it/v1/endpoints/CreateAndAmendReliefInvestmentsControllerISpec.scala:530]: "INVALID_CORRELATIONID" will probs need to be added here to the ISpec too

**Review by sean-flynn-HMRC — APPROVED** [review]
Very good

**Pull request merged**